File: packages/core/src/collector.ts
import fs from 'node:fs'
import path from 'node:path'
import os from 'node:os'
import type { Session, SessionFile } from './types.js'

const PROJECTS_DIR = path.join(os.homedir(), '.claude', 'projects')

/** 列出所有 ClaudeCode session 文件（仅 stat，不读内容） */
export function listSessionFiles(): SessionFile[] {
  const result: SessionFile[] = []
  if (!fs.existsSync(PROJECTS_DIR)) return result

  for (const projDirName of fs.readdirSync(PROJECTS_DIR)) {
    const projDir = path.join(PROJECTS_DIR, projDirName)
    let dirStat: fs.Stats
    try {
      dirStat = fs.statSync(projDir)
    } catch {
      continue
    }
    if (!dirStat.isDirectory()) continue

    for (const fileName of fs.readdirSync(projDir)) {
      if (!fileName.endsWith('.jsonl')) continue
      const filePath = path.join(projDir, fileName)
      let fstat: fs.Stats
      try {
        fstat = fs.statSync(filePath)
      } catch {
        continue
      }
      result.push({
        filePath,
        fileName,
        sessionId: extractSessionId(fileName),
        fileSizeBytes: fstat.size,
        mtime: fstat.mtimeMs
      })
    }
  }
  return result
}

function extractSessionId(fileName: string): string | null {
  const base = fileName.replace(/\.jsonl$/, '')
  const m = base.match(/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/i)
  return m ? m[0] : null
}

interface SessionEvent {
  type?: string
  sessionId?: string
  cwd?: string
  gitBranch?: string
  version?: string
  timestamp?: string
  isMeta?: boolean
  subtype?: string
  level?: string
  message?: {
    role?: string
    model?: string
    content?: unknown
  }
}

/** 解析失败的原因：
 *  - 'empty'：文件为空或不可读
 *  - 'no-conversation'：文件存在但无对话内容（Claude Code 写的 summary/snapshot 辅助文件）
 *  - 'read-error'：读取抛异常
 */
export type ParseFailureReason = 'empty' | 'no-conversation' | 'read-error'

export type ParseResult = { ok: true; session: Session } | { ok: false; reason: ParseFailureReason }

/** 全量解析一个 session 文件，返回结构化 Session 或失败原因 */
export function parseSessionFileResult(filePath: string): ParseResult {
  let content: string
  try {
    content = fs.readFileSync(filePath, 'utf8')
  } catch {
    return { ok: false, reason: 'read-error' }
  }
  const lines = content.split('\n').filter(Boolean)
  if (lines.length === 0) return { ok: false, reason: 'empty' }

  let sessionId: string | null = null
  let projectPath = ''
  let gitBranch: string | null = null
  let claudeVersion: string | null = null
  let startedAt: number | null = null
  let endedAt: number | null = null
  let userMessageCount = 0
  let assistantMessageCount = 0
  let toolCallCount = 0
  let errorCount = 0
  const models = new Set<string>()

  for (const line of lines) {
    let event: SessionEvent
    try {
      event = JSON.parse(line) as SessionEvent
    } catch {
      continue
    }

    if (event.sessionId && !sessionId) sessionId = event.sessionId
    if (event.cwd) projectPath = event.cwd
    if (event.gitBranch) gitBranch = event.gitBranch
    if (event.version) claudeVersion = event.version
    if (typeof event.timestamp === 'string') {
      const ts = new Date(event.timestamp).getTime()
      if (!Number.isNaN(ts)) {
        if (startedAt === null || ts < startedAt) startedAt = ts
        if (endedAt === null || ts > endedAt) endedAt = ts
      }
    }

    if (event.type === 'user') {
      if (!event.isMeta) userMessageCount++
      const c = event.message?.content
      if (Array.isArray(c)) {
        for (const block of c) {
          if (block && typeof block === 'object' && (block as { type?: string }).type === 'tool_result') {
            toolCallCount++
          }
        }
      }
    } else if (event.type === 'assistant') {
      assistantMessageCount++
      if (event.message?.model) models.add(event.message.model)
      const c = event.message?.content
      if (Array.isArray(c)) {
        for (const block of c) {
          if (block && typeof block === 'object' && (block as { type?: string }).type === 'tool_use') {
            toolCallCount++
          }
        }
      }
    } else if (event.type === 'system') {
      if (event.subtype === 'api_error' || event.level === 'error') errorCount++
    }
  }

  // 没在内容里找到 sessionId：fallback 用文件名里的 uuid
  // （summary / file-history-snapshot 这类辅助文件没有 sessionId 字段，但文件名遵循 uuid 命名）
  if (!sessionId) {
    sessionId = extractSessionId(path.basename(filePath))
  }
  // 既没 sessionId 又没任何对话内容（user/assistant）：归类为「无对话」而非「失败」
  if (!sessionId || (userMessageCount === 0 && assistantMessageCount === 0)) {
    return { ok: false, reason: 'no-conversation' }
  }

  const stat = fs.statSync(filePath)
  const now = Date.now()
  const start = startedAt ?? now
  const end = endedAt ?? start
  const durationMs = Math.max(0, end - start)

  return {
    ok: true,
    session: {
      id: sessionId,
      filePath,
      projectPath,
      gitBranch,
      claudeVersion,
      startedAt: start,
      endedAt: end,
      durationMs,
      userMessageCount,
      assistantMessageCount,
      toolCallCount,
      errorCount,
      models: [...models],
      fileSizeBytes: stat.size,
      lineCount: lines.length,
      lastParsedLineCount: lines.length,
      ingestedAt: now
    }
  }
}

/** 兼容旧调用：返回 Session 或 null（不区分失败原因） */
export function parseSessionFile(filePath: string): Session | null {
  const r = parseSessionFileResult(filePath)
  return r.ok ? r.session : null
}

File: apps/desktop/src-tauri/src/collector.rs
//! 对应 @navi/core/src/collector.ts：扫描 ~/.claude/projects 下的 session jsonl 并解析

use crate::util::parse_js_date;
use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionFile {
    pub file_path: String,
    pub file_name: String,
    pub session_id: Option<String>,
    pub file_size_bytes: i64,
    pub mtime: f64,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Session {
    pub id: String,
    pub file_path: String,
    pub project_path: String,
    pub git_branch: Option<String>,
    pub claude_version: Option<String>,
    pub started_at: i64,
    pub ended_at: i64,
    pub duration_ms: i64,
    pub user_message_count: i64,
    pub assistant_message_count: i64,
    pub tool_call_count: i64,
    pub error_count: i64,
    pub models: Vec<String>,
    pub file_size_bytes: i64,
    pub line_count: i64,
    pub last_parsed_line_count: i64,
    pub ingested_at: i64,
}

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum ParseFailureReason {
    Empty,
    NoConversation,
    ReadError,
}

pub enum ParseResult {
    Ok(Session),
    Fail(ParseFailureReason),
}

fn projects_dir() -> std::path::PathBuf {
    dirs::home_dir().unwrap_or_default().join(".claude").join("projects")
}

fn extract_session_id(file_name: &str) -> Option<String> {
    let base = file_name.strip_suffix(".jsonl").unwrap_or(file_name);
    let re = regex::Regex::new(r"(?i)[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}").unwrap();
    re.find(base).map(|m| m.as_str().to_string())
}

/// 列出所有 ClaudeCode session 文件（仅 stat，不读内容）
pub fn list_session_files() -> Vec<SessionFile> {
    let mut result = Vec::new();
    let root = projects_dir();
    if !root.exists() {
        return result;
    }
    let proj_dirs = match std::fs::read_dir(&root) {
        Ok(d) => d,
        Err(_) => return result,
    };
    for entry in proj_dirs.flatten() {
        let proj_dir = entry.path();
        if !proj_dir.is_dir() {
            continue;
        }
        let files = match std::fs::read_dir(&proj_dir) {
            Ok(d) => d,
            Err(_) => continue,
        };
        for f in files.flatten() {
            let file_name = f.file_name().to_string_lossy().to_string();
            if !file_name.ends_with(".jsonl") {
                continue;
            }
            let file_path = f.path();
            let md = match f.metadata() {
                Ok(m) => m,
                Err(_) => continue,
            };
            let mtime = md
                .modified()
                .ok()
                .and_then(|t| t.duration_since(std::time::UNIX_EPOCH).ok())
                .map(|d| d.as_millis() as f64)
                .unwrap_or(0.0);
            result.push(SessionFile {
                session_id: extract_session_id(&file_name),
                file_path: file_path.to_string_lossy().to_string(),
                file_name,
                file_size_bytes: md.len() as i64,
                mtime,
            });
        }
    }
    result
}

/// 全量解析一个 session 文件，返回结构化 Session 或失败原因
pub fn parse_session_file_result(file_path: &str) -> ParseResult {
    let content = match std::fs::read_to_string(file_path) {
        Ok(c) => c,
        Err(_) => return ParseResult::Fail(ParseFailureReason::ReadError),
    };
    let lines: Vec<&str> = content.split('\n').filter(|l| !l.is_empty()).collect();
    if lines.is_empty() {
        return ParseResult::Fail(ParseFailureReason::Empty);
    }

    let mut session_id: Option<String> = None;
    let mut project_path = String::new();
    let mut git_branch: Option<String> = None;
    let mut claude_version: Option<String> = None;
    let mut started_at: Option<i64> = None;
    let mut ended_at: Option<i64> = None;
    let mut user_message_count: i64 = 0;
    let mut assistant_message_count: i64 = 0;
    let mut tool_call_count: i64 = 0;
    let mut error_count: i64 = 0;
    let mut models: Vec<String> = Vec::new();

    for line in &lines {
        let ev: serde_json::Value = match serde_json::from_str(line) {
            Ok(v) => v,
            Err(_) => continue,
        };

        // JS 语义：truthy（非空串）才赋值；sessionId 取首个，cwd/gitBranch/version 取最后一个非空
        if session_id.is_none() {
            if let Some(sid) = ev.get("sessionId").and_then(|v| v.as_str()).filter(|s| !s.is_empty()) {
                session_id = Some(sid.to_string());
            }
        }
        if let Some(cwd) = ev.get("cwd").and_then(|v| v.as_str()).filter(|s| !s.is_empty()) {
            project_path = cwd.to_string();
        }
        if let Some(b) = ev.get("gitBranch").and_then(|v| v.as_str()).filter(|s| !s.is_empty()) {
            git_branch = Some(b.to_string());
        }
        if let Some(v) = ev.get("version").and_then(|v| v.as_str()) {
            if !v.is_empty() {
                claude_version = Some(v.to_string());
            }
        }
        if let Some(ts) = ev.get("timestamp").and_then(|v| v.as_str()) {
            if let Some(t) = parse_js_date(ts) {
                if started_at.is_none() || t < started_at.unwrap() {
                    started_at = Some(t);
                }
                if ended_at.is_none() || t > ended_at.unwrap() {
                    ended_at = Some(t);
                }
            }
        }

        let ev_type = ev.get("type").and_then(|v| v.as_str()).unwrap_or("");
        let is_meta = ev.get("isMeta").and_then(|v| v.as_bool()).unwrap_or(false);
        match ev_type {
            "user" => {
                if !is_meta {
                    user_message_count += 1;
                }
                if let Some(c) = ev.pointer("/message/content").and_then(|v| v.as_array()) {
                    for block in c {
                        if block.get("type").and_then(|t| t.as_str()) == Some("tool_result") {
                            tool_call_count += 1;
                        }
                    }
                }
            }
            "assistant" => {
                assistant_message_count += 1;
                if let Some(m) = ev.pointer("/message/model").and_then(|v| v.as_str()) {
                    if !models.iter().any(|x| x == m) {
                        models.push(m.to_string());
                    }
                }
                if let Some(c) = ev.pointer("/message/content").and_then(|v| v.as_array()) {
                    for block in c {
                        if block.get("type").and_then(|t| t.as_str()) == Some("tool_use") {
                            tool_call_count += 1;
                        }
                    }
                }
            }
            "system" => {
                let subtype = ev.get("subtype").and_then(|v| v.as_str());
                let level = ev.get("level").and_then(|v| v.as_str());
                if subtype == Some("api_error") || level == Some("error") {
                    error_count += 1;
                }
            }
            _ => {}
        }
    }

    // 没在内容里找到 sessionId：fallback 用文件名里的 uuid
    if session_id.is_none() {
        session_id = extract_session_id(&crate::util::basename(file_path));
    }
    if session_id.is_none() || (user_message_count == 0 && assistant_message_count == 0) {
        return ParseResult::Fail(ParseFailureReason::NoConversation);
    }

    let stat_size = std::fs::metadata(file_path).map(|m| m.len() as i64).unwrap_or(0);
    let now = crate::paths::now_ms();
    let start = started_at.unwrap_or(now);
    let end = ended_at.unwrap_or(start);
    let duration_ms = (end - start).max(0);

    ParseResult::Ok(Session {
        id: session_id.unwrap(),
        file_path: file_path.to_string(),
        project_path,
        git_branch,
        claude_version,
        started_at: start,
        ended_at: end,
        duration_ms,
        user_message_count,
        assistant_message_count,
        tool_call_count,
        error_count,
        models,
        file_size_bytes: stat_size,
        line_count: lines.len() as i64,
        last_parsed_line_count: lines.len() as i64,
        ingested_at: now,
    })
}
